### ISSUE TITLE
[Feature] Add switch to is_bearable and die_if_unbearable to fail is beartype can't type-check the type completely

### ISSUE THREAD

**Issue by gabrieldemarmiesse** [beartype/beartype#279]
Hi there! 

A bit of background... I became fan of beartype, then I had to look at beartype related projects. Found https://github.com/beartype/plum and loved it. Made a few contributions, then decided that a new lib was needed to have full compatibility between `@typing.overload` and a multiple dispatch library (backward compatibility issues and all that stuff...). 
Beartype has a new baby bear and I'm quite proud of it: https://github.com/gabrieldemarmiesse/overtake .  This little bear is making everyone happy in bearland, including our sworn ennemie, the static type-checkers.

But I disgress. Let's come back to the topic at hand. I'm using `is_bearable` and `die_if_unbearable`. But then I'm having this issue:

```python
from overtake import overtake
from typing import overload


@overload
def what_is_this(obj: dict[str, str]) -> None:
    print("This is a dict[str, str]")


@overload
def what_is_this(obj: dict[int, int]) -> None:
    print("This is a dict[int, int]")


@overtake
def what_is_this(obj):
    ...


what_is_this({1: 1})
# This is a dict[str, str]
```

The users won't be happy at all.
The bear might be roaring and might tell me: "What are you going on about? There is already an issue about it: https://github.com/beartype/beartype/issues/167" but that's not the feature I'm asking.

I'd prefer to have something along the lines of "This type is not supported by beartype yet. Come back later!".

This is because `is_bearable({1: 1}, dict[str, str]) is True`. While that makes sense with the `@beartype` decorator, for other purposes like multiple dispatch, having a silent error is quite surprising.

We can't expect beartype to be able to handle all types in the world, especially the new ones, but we can expect beartype to tell us if a type can't be completely checked.

What I would propose is a new argument so that users can write `is_bearable({1:1}, dict[str, str], error_if_unknown=True)` and get a beartype exception is beartype can't handle the type. Same with `die_if_unbearable`. Default would be `False` for this flag, for backward compatiblity reasons of course. 

I believe `plum` can also benefit from such a flag, that would avoid silent errors.

**Comment by leycec** [comment]
>  Beartype has a new baby bear and I'm quite proud of it: https://github.com/gabrieldemarmiesse/overtake .

**So fun!** Rascally baby bear, @beartype welcomes you into the world. When hungry, please mewl, roll over onto your belly, and beg Pappa @gabrieldemarmiesse to feed you more ~~blueberries coated in honey~~ commits. Which reminds me...

@wesselb: I couldn't help but notice while stalking Plum's commit history that you recently implemented something similar for Plum as `@plum.overload`. How does `@overtake.overtake` compare to `@plum.overload`? Their use pattern appears broadly similar – but the horned devil is always in the details. Should @beartype publish a new FAQ entry on multiple dispatch comparing and contrasting the merits of `@plum.overload`, `@overtake.overtake`, and `@typing.overload`? If so, jam that emoji button until @leycec finally does something.

Back to the exciting and thought-provoking topic at hand that makes me stroke my beard stubble...

> I'd prefer to have something along the lines of "This type is not supported by beartype yet. Come back later!".

**Reasonable!** Very reasonable. Certainly, we could and possibly should do this. The main issue with doing anything, though, is always...

:watch: :timer_clock: :hourglass: 

*...time.* @beartype's age-old arch nemesis, come to bedevil us again with its petty mortality, insipid entropy, and causal inexorability. Curse you, Father Time, you callous fiend! We are reminded at this juncture of [classic 90's German power metal that now makes me feel old and sad](https://www.youtube.com/watch?v=Dr9dVHtgJ7M):

> Time, what is time?
> I wish I knew how to tell you why.
> It hurts to know.
> Aren't we machines?
> Time, what is time?
> Unlock the door
> And see the truth.
> Then time is time again.

I'm pretty sure that's a thinly veiled reference to *Bladerunner*. What isn't, though? :face_exhaling: 

In short, doing this would take time. In fact, doing this might take even more time than just deeply type-checking `dict[...]` type hints already. Everyone wants that. So, I should probably just do that.

Let's see if @leycec does anything at all. Only old-guy Harrison Ford knows the truth.

**Comment by gabrieldemarmiesse** [comment]
> In short, doing this would take time. In fact, doing this might take even more time than just deeply type-checking dict[...] type hints already. Everyone wants that. So, I should probably just do that.

I'm not sure that solves the real problem though. If there are new types that beartype can't deep-check yet then there should be an option to know about it automatically, otherwise users might think `is_bearable()` is doing its job, when it's actually not. But again, it's just an idea, and feel free to place it in your agenda wherever you want :)

**Comment by wesselb** [comment]
> @wesselb: I couldn't help but notice while stalking Plum's commit history that you recently implemented something similar for Plum as @plum.overload. How does @overtake.overtake compare to @plum.overload? Their use pattern appears broadly similar – but the horned devil is always in the details. Should https://github.com/beartype publish a new FAQ entry on multiple dispatch comparing and contrasting the merits of @plum.overload, @overtake.overtake, and @typing.overload? If so, jam that emoji button until @leycec finally does something.

Yes, they are indeed super similar! `plum.overload` was actually developed together with @gabrieldemarmiesse. :) @gabrieldemarmiesse, please correct me if I'm wrong, but I think that the main difference is how keyword arguments are handled. In Plum, keyword arguments are ignored in the dispatch process, whereas Overtake does take them into account. An FAQ entry would be super awesome, though I think that the judge is still out on whether the `overload`-based pattern to do multiple dispatch in a `mypy`-compliant way really is the way to go, as it has a number of unfortunate limitations.

**Comment by leycec** [comment]
> If there are new types that @beartype can't deep-check yet then there should be an option to know about it automatically, otherwise users might think `is_bearable()` is doing its job, when it's actually not.

You're not wrong, @gabrieldemarmiesse. You're actually right. Again, the issue is time; this is yet another feature request for which no relevant implementation, infrastructure, or API exists. Off the top of my bald head, conveying this metadata would require @beartype to additionally:

* Internally maintain a frozen set listing all kinds of type hints that @beartype only shallowly type-checks but *should* deeply type-check. Since this set doesn't exist yet, I'd need to parse this set from [our ReadTheDocs (RTD)-hosted documentation](https://beartype.readthedocs.io/en/latest/pep). I feel tired and old already.
* Add a new optional keyword-only parameter to our public `beartype.BeartypeConf` dataclass enabling users to conditionally enable this reporting feature (e.g., as `BeartypeConf(is_warn_hint_deeply_unsupported=True)` or *something something*).
* Add a new public `beartype.roar.BeartypeDecorHintDeeplyUnsupported` warning category.
* Refactor @beartype's code generation algorithm. Specifically, for the currently visited type hint at decoration-time (i.e., code generation time), @beartype should now:
  1. Detect whether `bear_call.conf.is_warn_hint_deeply_unsupported` is `True`.
  1. If so, detect whether the current type hint is in the frozen set described above.
  1. If so, emit a non-fatal `BeartypeDecorHintDeeplyUnsupported` warning.
* Exhaustively document that new parameter.
* Exhaustively unit test that new parameter.
* Lastly, whenever we implement deep type-checking support for a new kind of type hint, we must remember to **remove that type hint from the frozen set described above.** In other words, we are now inviting desynchronization woes. If we forget to do something, @beartype starts emitting false positives. That's bad.

This is all feasible – but it's also a surprisingly non-trivial amount of work. This is why I suggested that just implementing deep type-checking support for common containers that everybody wants like `dict[..., ...]` and `set[...]` is probably easier. We don't need to add *any* new parameters, warnings, documentation, or even tests to do that. When we deeply type-check something new, our public API doesn't need to change; only our internal code generation algorithm needs to change.

In fact, I could probably have just started implementing deep type-checking support for this stuff instead of wastefully typing the above dissertation that nobody justifiably cares about. It is sad! Cry, cat emoji! :crying_cat_face:

**Comment by gabrieldemarmiesse** [comment]
That makes sense, thanks!
